### ISSUE TITLE
dask.multiprocessing.KeyError

### ISSUE THREAD

**Issue by baranberkay96** [aeturrell/occupationcoder#13]
* occupationcoder version: 0.2.0
* Python version: Python 3.7.7
* Operating System: MacOS BigSur 11.2.3
* pip freeze output:

```
beautifulsoup4==4.9.3
click==7.1.2
cloudpickle==1.6.0
dask==2021.4.0
fsspec==2021.4.0
joblib==1.0.1
locket==0.2.1
nltk==3.6.2
numpy==1.20.2
occupationcoder==0.2.0
pandas==1.2.4
partd==1.2.0
python-dateutil==2.8.1
pytz==2021.1
PyYAML==5.4.1
regex==2021.4.4
scikit-learn==0.24.1
scipy==1.6.2
six==1.15.0
soupsieve==2.2.1
threadpoolctl==2.1.0
toolz==0.11.1
tqdm==4.60.0
```

### Description
Hi @aeturrell, we are working with the latest python3 update. We have the sample data: 
- job_title: Ground worker
- job_description: Options Resourcing are seeking an experienced ground worker for a project that will Start in Chippenham then move to other local parts of the west country
- job_sector: Construction

We want to run this snippet in python environment: 

```
import pandas as pd
from occupationcoder.coder import coder
myCoder = coder.Coder()

if __name__ == '__main__':
    myCoder.codejobrow('Ground worker','Options Resourcing are seeking an experienced ground worker for a project that will Start in Chippenham then move to other local parts of the west country','Construction')
```

### What I Did
But we are getting this error, could not solve. Any help would be pretty good, thanks :)
```
Traceback (most recent call last):
  File "analyze.py", line 19, in <module>
    myCoder.codejobrow('Ground worker','experienced ground worker','Construction')
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/occupationcoder/coder/coder.py", line 42, in codejobrow
    return self.codedataframe(df)
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/occupationcoder/coder/coder.py", line 98, in codedataframe
    x = res.compute(scheduler='processes')
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/dask/base.py", line 284, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/dask/base.py", line 566, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/dask/multiprocessing.py", line 230, in get
    **kwargs
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/dask/local.py", line 514, in get_async
    raise_exception(exc, tb)
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/dask/multiprocessing.py", line 109, in reraise
    raise exc
dask.multiprocessing.KeyError: 814

Traceback
---------
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/dask/local.py", line 223, in execute_task
    result = _execute_task(task, data)
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/dask/core.py", line 121, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/dask/optimization.py", line 963, in __call__
    return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/dask/core.py", line 151, in get
    result = _execute_task(task, cache)
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/dask/core.py", line 121, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/dask/utils.py", line 34, in apply
    return func(*args, **kwargs)
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/dask/dataframe/core.py", line 5665, in apply_and_enforce
    df = func(*args, **kwargs)
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/dask/utils.py", line 900, in __call__
    return getattr(obj, self.method)(*args, **kwargs)
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/pandas/core/frame.py", line 7768, in apply
    return op.get_result()
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/pandas/core/apply.py", line 185, in get_result
    return self.apply_standard()
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/pandas/core/apply.py", line 276, in apply_standard
    results, res_index = self.apply_series_generator()
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/pandas/core/apply.py", line 290, in apply_series_generator
    results[i] = self.f(v)
  File "/Users/baranberkaybarakcin/Documents/learning/occupation-coder/env/lib/python3.7/site-packages/occupationcoder/utilities/utilities.py", line 406, in return_best_match_2
    good_opt = extract(str(dataframe_row['title_nospace']), titles_mg[c])
```

**Comment by ollycohen** [comment]
Unfortunately, I don't have the solution, but I just want to say I've been getting the same error. I had been executing in a Jupyter notebook and then switched to a python script with the `if __name__`... statement. My error doesn't say it comes from dask. It just says,`KeyError: 115`, but still comes from line 406 in utilities.py. Thanks for sharing the error, and hope it gets solved.

**Comment by baranberkay96** [comment]
Hi, I found the problem and I solve it however, it requires tests. In utilities.py line 400: `codes = dataframe_row['top5']` give us a array with integers. In line 406: `good_opt = extract(str(dataframe_row['title_nospace']), titles_mg[c])` we want to select item from dictionary with c as integer. When I found that, I examine titles_mg and I saw that keys are string type. And I edited 406: as `good_opt = extract(str(dataframe_row['title_nospace']), titles_mg[str(c)])`  and it worked. Here is the output:

`       
job_title  ... SOC_code
0  Ground worker  ...      912
`

I am going to be very happy if it's correct. It would be my first contribution to open source :).

**Comment by aeturrell** [comment]
Thank you @ollycohen and especially @baranberkay96 for flagging this, and a huge thanks to @baranberkay96 for finding the issue. I have put a fix in now, see a96c6af. Note that this will not be available via pypi yet (ie you cannot fix with `pip install occupationcoder`) as I've not made a new release, but it *is* available by downloading the code directly from github and installing that way.

**Comment by ccomunello** [comment]
Hi @aeturrell ! I am getting a bug on occupationcoder/utilities/utilities.py line 319 

`
def getKey(item):
    """
    Takes any iterable as input.
    Returns tuple.
    Is used to specify order of argument importance, which can be used for
    sorting and gettin max using multiple criteria.
    >>> getKey(('registered nurse', 90, 4))
    (90, 4, 2)
    """
    return (item[1], item[2], len(item[0].split()))

`

I get tuple index problems, but code is apparently ok. Full Traceback:

`WARNING:root:Applied processor reduces input query to empty string, all comparisons will have score 0. [Query: 'â€“']
Traceback (most recent call last):
  File "./1adzuna.py", line 67, in <module>
    main()
  File "./1adzuna.py", line 57, in main
    df1=myCoder.codedataframe(d)
  File "/home/cc18002/.local/lib/python3.6/site-packages/occupationcoder/coder/coder.py", line 104, in codedataframe
    x = res.compute(scheduler='processes')
  File "/usr/local/lib/python3.6/site-packages/dask/base.py", line 156, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/dask/base.py", line 397, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/dask/multiprocessing.py", line 192, in get
    raise_exception=reraise, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/dask/local.py", line 501, in get_async
    raise_exception(exc, tb)
  File "/usr/local/lib/python3.6/site-packages/dask/compatibility.py", line 111, in reraise
    raise exc.with_traceback(tb)
  File "/usr/local/lib/python3.6/site-packages/dask/local.py", line 272, in execute_task
    result = _execute_task(task, data)
  File "/usr/local/lib/python3.6/site-packages/dask/local.py", line 253, in _execute_task
    return func(*args2)
  File "/usr/local/lib/python3.6/site-packages/dask/dataframe/core.py", line 3684, in apply_and_enforce
    df = func(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/dask/utils.py", line 694, in __call__
    return getattr(obj, self.method)(*args, **kwargs)
  File "/home/cc18002/.local/lib/python3.6/site-packages/pandas/core/frame.py", line 6928, in apply
    return op.get_result()
  File "/home/cc18002/.local/lib/python3.6/site-packages/pandas/core/apply.py", line 186, in get_result
    return self.apply_standard()
  File "/home/cc18002/.local/lib/python3.6/site-packages/pandas/core/apply.py", line 292, in apply_standard
    self.apply_series_generator()
  File "/home/cc18002/.local/lib/python3.6/site-packages/pandas/core/apply.py", line 321, in apply_series_generator
    results[i] = self.f(v)
  File "/home/cc18002/.local/lib/python3.6/site-packages/occupationcoder/utilities/utilities.py", line 392, in return_best_match_2
    final_code = max(items(), key=lambda x: getKey(x[1]))
  File "/home/cc18002/.local/lib/python3.6/site-packages/occupationcoder/utilities/utilities.py", line 392, in <lambda>
    final_code = max(items(), key=lambda x: getKey(x[1]))
  File "/home/cc18002/.local/lib/python3.6/site-packages/occupationcoder/utilities/utilities.py", line 319, in getKey
    return (item[1], item[2], len(item[0].split()))
IndexError: ('tuple index out of range', 'occurred at index 706')
`


The problem seems to be in a specific data file, but looking over the file structure, variable class, string text etc, there is no apparent difference from other files. The code I am running is below, any guesses of where it might be coming from? Thanks!!!

`#!/usr/bin/python36

import pandas as pd
from occupationcoder.coder import coder
import os.path
import re
import os
from os import walk
import glob
import nltk
nltk.download('punkt')	
nltk.download('wordnet')
nltk.download('omw-1.4')


dir = '/home/ALL/'
dir1 = '/home/

myCoder = coder.Coder()


d0 = [[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]]

df0 = pd.DataFrame(data=d0, columns=[ 'year', 'month', 'day', 'category_id', 'company', 'date_created','location_raw', 'salary_max', 'salary_min', 'salary_predicted',
       'salary_currency', 'salary_raw', 'title', 'category_name', 'location_path', 'contract_time',
       'contract_type', 'company_id', 'company_name', 'description'] )
df0.to_csv(os.path.join(dir1, 'vacancy_stock_raw.csv'))
df0=pd.read_csv(os.path.join(dir1, 'vacancy_stock_raw.csv'))


#loop over folders and append selected variables into vacancy_stock_raw.csv
def main():
    df2=df0
    
    li=[]
    r = []
    def list_files(dir):
        r = []
    for root, dirs, files in os.walk(dir):
        for name in files:
            filepath = root + os.sep + name
            if filepath.endswith(".csv"):
                print(filepath)
                d = pd.read_csv(filepath, index_col=None, header=0)
                d = d.rename(columns={'title': 'job_title'})
                d = d.rename(columns={'description': 'job_description'})
                d = d.rename(columns={'category_name': 'job_sector'})
                d=d.astype(str)
                d['job_title'] = d['job_title'].str.slice(0, 100)
                d['job_sector'] = d['job_sector'].str.slice(0, 100)
                d['job_description'] = d['job_description'].str.slice(0, 200)
                d = d[[ 'year', 'month', 'day', 'category_id', 'company', 'date_created','location_raw', 'salary_max', 'salary_min', 'salary_predicted',
           'salary_currency', 'job_title', 'job_sector', 'location_path', 'contract_time',
           'contract_type', 'company_id', 'company_name', 'job_description']]
                d['titleno_space'] = d['job_title']
                print('ready')
                df1=myCoder.codedataframe(d)
                print('socdone')
                df1 = df1[[ 'year', 'month', 'day', 'category_id', 'company', 'date_created','location_raw', 'salary_max', 'salary_min', 'salary_predicted',
           'salary_currency', 'job_title', 'job_sector', 'location_path', 'contract_time',
           'contract_type', 'company_id', 'company_name', 'SOC_code']]
                df2=df2.append(df1)
                df2.to_csv(os.path.join(dir1, 'vacancy_stock_raw.csv'))

                
if __name__ == '__main__':
    main()
`
